crates/ibc/src/core/ics04_channel/handler/chan_close_confirm.rs

```diff
@@ -3,82 +3,102 @@ use crate::core::ics03_connection::connection::State as ConnectionState;
 use crate::core::ics04_channel::channel::{ChannelEnd, Counterparty, State};
 use crate::core::ics04_channel::context::ChannelReader;
 use crate::core::ics04_channel::error::Error;
-use crate::core::ics04_channel::handler::verify::verify_channel_proofs;
 use crate::core::ics04_channel::handler::{ChannelIdState, ChannelResult};
 use crate::core::ics04_channel::msgs::chan_close_confirm::MsgChannelCloseConfirm;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
+/// Per our convention, this message is processed on chain B.
 pub(crate) fn process<Ctx: ChannelReader>(
-    ctx: &Ctx,
+    ctx_b: &Ctx,
     msg: &MsgChannelCloseConfirm,
 ) -> HandlerResult<ChannelResult, Error> {
     let mut output = HandlerOutput::builder();
 
     // Retrieve the old channel end and validate it against the message.
-    let mut channel_end = ctx.channel_end(&msg.port_id, &msg.channel_id)?;
+    let chan_end_on_b = ctx_b.channel_end(&msg.port_id_on_b, &msg.chan_id_on_b)?;
 
     // Validate that the channel end is in a state where it can be closed.
-    if channel_end.state_matches(&State::Closed) {
-        return Err(Error::channel_closed(msg.channel_id.clone()));
+    if chan_end_on_b.state_matches(&State::Closed) {
+        return Err(Error::channel_closed(msg.chan_id_on_b.clone()));
     }
 
     // An OPEN IBC connection running on the local (host) chain should exist.
-    if channel_end.connection_hops().len() != 1 {
+    if chan_end_on_b.connection_hops().len() != 1 {
```

Comment:
I think we can move stateless checks like these into the `TryFrom<RawMsg>` conversion.

Reply:
Actually, there are a bunch of checks in ibc-go that we don't have in our handlers -> https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/msgs.go
For e.g. ibc-go fails a `MsgChanOpenInit` with whose state isn't `Init`, etc.

Reply:
> I think we can move stateless checks like these into the TryFrom<RawMsg> conversion.

This `chan_end_on_b` comes from the context though, so no `TryFrom<Raw>`.

Also in general I prefer having all the validation in the same place. Otherwise it's hard to know what are all the checks that are being applied. We can discuss in the issue you'll open.